PackTask throws NullReferenceException when NuspecProperties is specified

As soon as any NuspecProperties is specified, dotnet pack fails.

```
dotnet new classlib
dotnet restore
dotnet build
dotnet pack
```

The pack succeeds. Now modify .csproj by adding this property definition in the `PropertyGroup`:

```xml
    <NuspecProperties>$(NuspecProperties);GitCommitIdShort=70f1d9353d</NuspecProperties>
```

Then pack again:

```
dotnet pack

Microsoft (R) Build Engine version 15.1.546.27743
Copyright (C) Microsoft Corporation. All rights reserved.

  d -> C:\temp\d\bin\Debug\netstandard1.4\d.dll
C:\Program Files\dotnet\sdk\1.0.0-rc4-004812\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(104,5): error MSB4018: The "PackTask" task failed unexpectedly.\r [C:\temp\d\d.csproj]
C:\Program Files\dotnet\sdk\1.0.0-rc4-004812\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(104,5): error MSB4018: System.NullReferenceException: Object reference not set to an instance of an object.\r [C:\temp\d\d.csproj]
C:\Program Files\dotnet\sdk\1.0.0-rc4-004812\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(104,5): error MSB4018:    at NuGet.Commands.PackCommandRunner.BuildFromProjectFile(String path)\r [C:\temp\d\d.csproj]
C:\Program Files\dotnet\sdk\1.0.0-rc4-004812\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(104,5): error MSB4018:    at NuGet.Commands.PackCommandRunner.BuildPackage(String path)\r [C:\temp\d\d.csproj]
C:\Program Files\dotnet\sdk\1.0.0-rc4-004812\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(104,5): error MSB4018:    at NuGet.Commands.PackCommandRunner.BuildPackage()\r [C:\temp\d\d.csproj]
C:\Program Files\dotnet\sdk\1.0.0-rc4-004812\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(104,5): error MSB4018:    at NuGet.Build.Tasks.Pack.PackTask.Execute()\r [C:\temp\d\d.csproj]
C:\Program Files\dotnet\sdk\1.0.0-rc4-004812\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(104,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()\r [C:\temp\d\d.csproj]
C:\Program Files\dotnet\sdk\1.0.0-rc4-004812\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(104,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__25.MoveNext() [C:\temp\d\d.csproj]
```

```
dotnet --version
1.0.0-rc4-004812
```

--- a/src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets
+++ b/src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets
@@ -274,7 +274,9 @@ Copyright (c) .NET Foundation. All rights reserved.
     </PropertyGroup>
   </Target>
   
-  <Target Name="_GetProjectReferenceVersions" DependsOnTargets="_CalculateInputsOutputsForPack;$(GetPackageVersionDependsOn)">
+  <Target Name="_GetProjectReferenceVersions" 
+          Condition="'$(NuspecFile)' == ''"
+          DependsOnTargets="_CalculateInputsOutputsForPack;$(GetPackageVersionDependsOn)">
     <ConvertToAbsolutePath Paths="$(RestoreOutputPath)">
       <Output TaskParameter="AbsolutePaths" PropertyName="RestoreOutputAbsolutePath" />
     </ConvertToAbsolutePath>

--- a/src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs
+++ b/src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTaskLogic.cs
@@ -33,17 +33,7 @@ namespace NuGet.Build.Tasks.Pack
                 BasePath = request.NuspecBasePath,
                 NoPackageAnalysis = request.NoPackageAnalysis,
                 WarningProperties = WarningProperties.GetWarningProperties(request.TreatWarningsAsErrors, request.WarningsAsErrors, request.NoWarn),
-                PackTargetArgs = new MSBuildPackTargetArgs
-                {
-                    AllowedOutputExtensionsInPackageBuildOutputFolder = InitOutputExtensions(request.AllowedOutputExtensionsInPackageBuildOutputFolder),
-                    AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder = InitOutputExtensions(request.AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder),
-                    TargetPathsToAssemblies = InitLibFiles(request.BuildOutputInPackage),
-                    TargetPathsToSymbols = InitLibFiles(request.TargetPathsToSymbols),
-                    AssemblyName = request.AssemblyName,
-                    IncludeBuildOutput = request.IncludeBuildOutput,
-                    BuildOutputFolder = request.BuildOutputFolder,
-                    TargetFrameworks = ParseFrameworks(request)
-                }
+                PackTargetArgs = new MSBuildPackTargetArgs()
             };
 
             packArgs.Logger = new PackCollectorLogger(request.Logger, packArgs.WarningProperties);
@@ -62,28 +52,43 @@ namespace NuGet.Build.Tasks.Pack
                 packArgs.MinClientVersion = version;
             }
 
-            if (request.NuspecProperties != null && request.NuspecProperties.Any())
-            {
-                packArgs.Properties.AddRange(ParsePropertiesAsDictionary(request.NuspecProperties));
-                if (packArgs.Properties.ContainsKey("version"))
-                {
-                    packArgs.Version = packArgs.Properties["version"];
-                }
-            }
 
             InitCurrentDirectoryAndFileName(request, packArgs);
             InitNuspecOutputPath(request, packArgs);
-
-            if (request.IncludeSource)
-            {
-                packArgs.PackTargetArgs.SourceFiles = GetSourceFiles(request, packArgs.CurrentDirectory);
-                packArgs.Symbols = request.IncludeSource;
-            }
-
             PackCommandRunner.SetupCurrentDirectory(packArgs);
 
-            var contentFiles = ProcessContentToIncludeInPackage(request, packArgs);
-            packArgs.PackTargetArgs.ContentFiles = contentFiles;
+            if (!string.IsNullOrEmpty(request.NuspecFile))
+            {
+                if (request.NuspecProperties != null && request.NuspecProperties.Any())
+                {
+                    packArgs.Properties.AddRange(ParsePropertiesAsDictionary(request.NuspecProperties));
+                    if (packArgs.Properties.ContainsKey("version"))
+                    {
+                        packArgs.Version = packArgs.Properties["version"];
+                    }
+                }
+            }
+            else
+            {
+                // This only needs to happen when packing via csproj, not nuspec.
+                packArgs.PackTargetArgs.AllowedOutputExtensionsInPackageBuildOutputFolder = InitOutputExtensions(request.AllowedOutputExtensionsInPackageBuildOutputFolder);
+                packArgs.PackTargetArgs.AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder = InitOutputExtensions(request.AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder);
+                packArgs.PackTargetArgs.TargetPathsToAssemblies = InitLibFiles(request.BuildOutputInPackage);
+                packArgs.PackTargetArgs.TargetPathsToSymbols = InitLibFiles(request.TargetPathsToSymbols);
+                packArgs.PackTargetArgs.AssemblyName = request.AssemblyName;
+                packArgs.PackTargetArgs.IncludeBuildOutput = request.IncludeBuildOutput;
+                packArgs.PackTargetArgs.BuildOutputFolder = request.BuildOutputFolder;
+                packArgs.PackTargetArgs.TargetFrameworks = ParseFrameworks(request);
+
+                if (request.IncludeSource)
+                {
+                    packArgs.PackTargetArgs.SourceFiles = GetSourceFiles(request, packArgs.CurrentDirectory);
+                    packArgs.Symbols = request.IncludeSource;
+                }
+
+                var contentFiles = ProcessContentToIncludeInPackage(request, packArgs);
+                packArgs.PackTargetArgs.ContentFiles = contentFiles;
+            }
 
             return packArgs;
         }

--- a/src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs
+++ b/src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs
@@ -69,7 +69,8 @@ namespace NuGet.Commands
                 MachineWideSettings = packArgs.MachineWideSettings,
                 Build = false,
                 PackTargetArgs = packArgs.PackTargetArgs,
-                Files = new HashSet<ManifestFile>()
+                Files = new HashSet<ManifestFile>(),
+                ProjectProperties = new Dictionary<string, string>()
             };
         }
 

--- a/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs
+++ b/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs
@@ -637,6 +637,68 @@ namespace Dotnet.Integration.Test
         }
 
         [PlatformFact(Platform.Windows)]
+        // This test is asserting that nuspec can be packed via dotnet.exe
+        // without having to specify IncludeBuildOutput=false when using the
+        // --no-build switch.
+        public void PackCommand_PackProject_PackNuspecWithoutBuild()
+        {
+            var nuspecFileContent = @"<?xml version=""1.0""?>
+<package>
+  <metadata>
+    <id>PackedFromNuspec</id>
+    <version>1.2.1</version>
+    <authors>Microsoft</authors>
+    <owners>NuGet</owners>
+    <description>This was packed from nuspec</description>
+  </metadata>
+  <files>
+    <file src=""abc.txt"" target=""CoreCLR/"" />
+  </files>
+</package>";
+            using (var testDirectory = TestDirectory.Create())
+            {
+                var projectName = "ClassLibrary1";
+                var workingDirectory = Path.Combine(testDirectory, projectName);
+                var projectFile = Path.Combine(workingDirectory, $"{projectName}.csproj");
+                // Act
+                msbuildFixture.CreateDotnetNewProject(testDirectory.Path, projectName, " classlib");
+                File.WriteAllText(Path.Combine(workingDirectory, "abc.nuspec"), nuspecFileContent);
+                File.WriteAllText(Path.Combine(workingDirectory, "abc.txt"), "sample text");
+
+                using (var stream = new FileStream(projectFile, FileMode.Open, FileAccess.ReadWrite))
+                {
+                    var xml = XDocument.Load(stream);
+                    ProjectFileUtils.AddProperty(xml, "NuspecFile", "abc.nuspec");
+                    ProjectFileUtils.WriteXmlToFile(xml, stream);
+                }
+
+                msbuildFixture.PackProject(workingDirectory, projectName,
+                    $"-o {workingDirectory} --no-build");
+
+                var nupkgPath = Path.Combine(workingDirectory, $"PackedFromNuspec.1.2.1.nupkg");
+                Assert.True(File.Exists(nupkgPath), "The output .nupkg is not in the expected place");
+
+                using (var nupkgReader = new PackageArchiveReader(nupkgPath))
+                {
+                    var nuspecReader = nupkgReader.NuspecReader;
+
+                    // Validate the .nuspec.
+                    Assert.Equal("PackedFromNuspec", nuspecReader.GetId());
+                    Assert.Equal("1.2.1", nuspecReader.GetVersion().ToFullString());
+                    Assert.Equal("Microsoft", nuspecReader.GetAuthors());
+                    Assert.Equal("NuGet", nuspecReader.GetOwners());
+                    Assert.Equal("This was packed from nuspec", nuspecReader.GetDescription());
+                    Assert.False(nuspecReader.GetRequireLicenseAcceptance());
+
+                    // Validate the assets.
+                    var libItems = nupkgReader.GetFiles("CoreCLR").ToArray();
+                    Assert.Equal("CoreCLR/abc.txt", libItems[0]);
+                }
+
+            }
+        }
+
+        [PlatformFact(Platform.Windows)]
         public void PackCommand_PackNewDefaultProject_InstallPackageToOutputPath()
         {
             using (var testDirectory = TestDirectory.Create())

--- a/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs
+++ b/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs
@@ -607,6 +607,36 @@ namespace Dotnet.Integration.Test
         }
 
         [PlatformFact(Platform.Windows)]
+        public void PackCommand_PackProject_EmptyNuspecFilePropertyWithNuspecProperties()
+        {
+            using (var testDirectory = TestDirectory.Create())
+            {
+                var projectName = "ClassLibrary1";
+                var workingDirectory = Path.Combine(testDirectory, projectName);
+                var projectFile = Path.Combine(workingDirectory, $"{projectName}.csproj");
+                // Act
+                msbuildFixture.CreateDotnetNewProject(testDirectory.Path, projectName, " classlib");
+
+                using (var stream = new FileStream(projectFile, FileMode.Open, FileAccess.ReadWrite))
+                {
+                    var xml = XDocument.Load(stream);
+                    ProjectFileUtils.AddProperty(xml, "NuspecProperties", "token1=value1");
+
+                    ProjectFileUtils.WriteXmlToFile(xml, stream);
+                }
+
+                msbuildFixture.RestoreProject(workingDirectory, projectName, string.Empty);
+                
+                msbuildFixture.PackProject(workingDirectory, projectName,
+                    $"-o {workingDirectory} /p:NuspecFile=");
+
+                var nupkgPath = Path.Combine(workingDirectory, $"ClassLibrary1.1.0.0.nupkg");
+                Assert.True(File.Exists(nupkgPath), "The output .nupkg is not in the expected place");
+
+            }
+        }
+
+        [PlatformFact(Platform.Windows)]
         public void PackCommand_PackNewDefaultProject_InstallPackageToOutputPath()
         {
             using (var testDirectory = TestDirectory.Create())

